### ISSUE TITLE
highlight: add configurable line breaks

### ISSUE THREAD

**Issue by Nuc1eoN** [McShelby/hugo-theme-relearn#169]
Hi,
I have noticed that your theme, by default, introduces line breaks in code blocks. I do not think this should be the default behaviour, and especially on smaller screens like mobile phones this can be a real problem. IMO code blocks are supposed to be shown in "vanilla" form and not be tinkered with.. in this case a horizontal scrollbar is needed and even desired.
Maybe if it's a one-liner that's "ok"... but in any other case it's just bad.

This is also apparent in your own example site, e.g. https://mcshelby.github.io/hugo-theme-relearn/cont/markdown/#blockquotes

However in our documentation (due to larger code sections) it is more apparent and annoying. It gets visually cluttered and defeats the purpose. Even worse if the code block is supposed to show some kind of graphical representation.

I have fixed this in my theme with:
```css
pre code {
    white-space: pre;
}
```

IMHO this should be the default behavior for code blocks! But if you happen to disagree I would even be fine with an optional setting for it, e.g. in `config.toml` ;)

**Comment by McShelby** [comment]
This is much a matter of taste and case.

In my use case I prefer it to show up completly without the need to use a scroll bar to view the whole picture.

I agree, that on mobile with very limited width space, this can look really horrible if the code is extensive. A media query could be preferable in this case.

If a scroll bar is used, it requires further tweaking for the print style. But how to handle extensive code in case of print? Render a non-usable scrollbar, cut it off, line breaks?

What if a user wants only certain code block behave that way? A configurable shortcode may be a good way to implement this.

As you see, this makes things more complicated as they are now.

If you want to change this for your installation, you could add the necessary style to the `custom-header.html` template, for now.

**Comment by Nuc1eoN** [comment]
Hi,
I see, yes it it a matter of taste and use-case then! I was not aware that some people might want to have it wrapped.

> If a scroll bar is used, it requires further tweaking for the print style. But how to handle extensive code in case of print? Render a non-usable scrollbar, cut it off, line breaks?

I am not sure what you mean by this frankly? Do you have an example maybe?

> As you see, this makes things more complicated as they are now.

I see what you mean. I propose not to make things too complicated for now. I just think a user should be able to define this globally in his `config.toml` (shortcodes to override it are nice, but can come later).
I have simply added this to my stylesheet and it works very nice:
```css
pre code {
    white-space: pre;
}
```
Unfortunately this causes one glitch (becasue apparently this type of usage has not been accounted for):
The "copy-to-clipboard" button scrolls with the content, instead of staying in a fixed position.

This is the only annoyance so far for me.

**Comment by McShelby** [comment]
> > If a scroll bar is used, it requires further tweaking for the print style. But how to handle extensive code in case of print? Render a non-usable scrollbar, cut it off, line breaks?
> 
> I am not sure what you mean by this frankly? Do you have an example maybe?

If you try to print out long lines, things get really funny like:

- adding another (or many depending on how long the code line is) sheet of paper with just the overflown code lines and almost no context (the idea of a reasonable printout is, that you can recover the order of sheets once you've dropped the whole pile)
- adding scrollbars to the printout, etc.

> Unfortunately this causes one glitch (becasue apparently this type of usage has not been accounted for): The "copy-to-clipboard" button scrolls with the content, instead of staying in a fixed position.

I would prefer a scrollbar on block level as in the [docsy theme[(https://www.docsy.dev/docs/adding-content/shortcodes/#blockscover).

**Comment by Nuc1eoN** [comment]
> If you try to print out long lines, things get really funny like:
> 
> - adding another (or many depending on how long the code line is) sheet of paper with just the overflown code lines and almost no context (the idea of a reasonable printout is, that you can recover the order of sheets once you've dropped the whole pile)
> - adding scrollbars to the printout, etc.

Really I don't understand... What you mean by print? Sheets of paper??
Long lines show up fine here:
![image](https://user-images.githubusercontent.com/2538022/152895757-19bc7083-be0a-4f13-9375-978eb33f3394.png)
Apart from the copy-to-clipboard button:
![image](https://user-images.githubusercontent.com/2538022/152895857-f0d07f1b-059d-4cc3-8cf8-512a3bbbffba.png)
... which scrolls with the output.

Maybe you are talking about some advanced features/shortcodes which I don't use?

**Comment by McShelby** [comment]
>  What you mean by print? Sheets of paper??

Yes. Paper printouts. I know, a concept from the past and not very environmental friendly but user tend to do all that weird stuff ;-)

**Comment by Nuc1eoN** [comment]
Paper is a different medium with different attributes. I think if somebody needs his/her documentation on paper he/she should convert it to pdf's. The markdown is all there.

Alternatively there could be a button top right of the code block to toggle line breaks. Although for me that button should be optional as I do not want my users to print out our documentation :)
If that is needed we will create a dedicated pdf version.

**Comment by McShelby** [comment]
For the moment I prefer, you are making this changes locally in your `custom-header.html` as I do not plan to change this in the near future.

**Comment by Nuc1eoN** [comment]
Yes understood. But why in the `custom-header.html`?
What about my css version?

**Comment by McShelby** [comment]
Do you know [Hugo's template lookup rules](https://gohugo.io/templates/lookup-order/)?

The theme calls the template `custom-header.html` in its HTML `<head>`. The `custom-header.html` that comes with the theme is plain empty. 

You can place your own `custom-header.html` file inside of your own projects `layouts/partials` directory and Hugo will use this instead of the empty distributed version.

By that you overwrite given theme CSS with your own definitions. The advantage is, you don't have to care, once you're updating the theme to a new version as your own `custom-header.html` stays untouched.

Where do you currently make all these CSS changes?

**Comment by Nuc1eoN** [comment]
Hey thanks, I know I make use of those partials, but I try to keep the modifications minimal because otherwise I do not benefit from the fixes made to the relearn theme, or I need to "port" the new html files, in order to receive eventual fixes (for example last time I needed to do it to receive the search bar fix :).
The only real modifications I have made to the partials, is the new redesigned top-bar, the footer text, a favicon, the logo and thats basically it :)

So most of my theme adjustments I keep in my own `/static/css/gobolinux-theme.css` file :)

**Comment by Nuc1eoN** [comment]
> The theme calls the template custom-header.html in its HTML <head>. The custom-header.html that comes with the theme is plain empty.
> You can place your own custom-header.html file inside of your own projects layouts/partials directory and Hugo will use this instead of the empty distributed version.

Hm oh ok Im not sure I new about `custom-header.html` in this way... the only use I have made of partials was to overwrite existing hmtl files.
I will check out the docs then!!
Although I still dont know if it is worth is fort his change, since it is one line of css. (But I havent yet fixed the copy-to-clipboard button)

**Comment by McShelby** [comment]
I would recommend you to do **all** changes in `custom-header.html`. Upgrading the theme will become much easier for you.

**Comment by McShelby** [comment]
For now, I've decided to keep things simple and close this issue. Maybe at a later point, if there is more demand, I'll make changes.

For now a reasonable workaround using `custom-header.html` is available.
'

**Comment by McShelby** [comment]
After #569 I've changed my mind about this.

The main reason is, that using the `highlight` shortcode with line numbers and wrapped code lines becomes really funny to look at (means unusable).

Nevertheless I have to think about how to implement it and what options may be allowed.

**Comment by McShelby** [comment]
@Nuc1eoN Support for wrapping vs. scrolling has landed. Be the first, to check out this hot new shtuff.

As I guessed above, things got a bit complicated but I'm pretty sure I've tackled it without introducing incompatiblities.

In your case you probably want to get rid of the wrapping using a global setting in your `config.toml`:

````toml
[params]
  highlightWrap = false
````

https://mcshelby.github.io/hugo-theme-relearn/shortcodes/highlight/index.html#optional-settings

I am happy for feedback!
